### PR TITLE
Add `str::{Split,RSplit,SplitN,RSplitN,SplitTerminator,RSplitTerminator,SplitInclusive}::as_str` methods

### PR DESCRIPTION
**Pull request by WaffleLapkin**
tl;dr this allows viewing unyelded part of str-split-iterators, like so:
```rust
let mut split = "Mary had a little lamb".split(' '); 
assert_eq!(split.as_str(), "Mary had a little lamb");
split.next();
assert_eq!(split.as_str(), "had a little lamb");
split.by_ref().for_each(drop);
assert_eq!(split.as_str(), "");
```

--------------

This PR adds semi-identical `as_str` methods to most str-split-iterators with signatures like `&'_ Split<'a, P: Pattern<'a>> -> &'a str` (Note: output `&str` lifetime is bound to the `'a`, not the `'_`). The methods are similar to [`Chars::as_str`]

`SplitInclusive::as_str` is under `"str_split_inclusive_as_str"` feature gate, all other methods are under `"str_split_as_str"` feature gate.

Before this PR you had to sum `len`s of all yielded parts or collect into `String` to emulate `as_str`.

[`Chars::as_str`]: https://doc.rust-lang.org/core/str/struct.Chars.html#method.as_str

### CODE REVIEW

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -1231,6 +1231,17 @@ impl<'a, P: Pattern<'a>> SplitInternal<'a, P> {
             },
         }
     }
+
+    #[inline]
+    fn as_str(&self) -> &'a str {
+        // `Self::get_end` doesn't change `self.start`
+        if self.finished {
+            return "";
```
Comment (crlf0710) [inline library/core/src/str/mod.rs:1239]: This will cause surprising result for those who want to calculate ptr offsets of this &str from the whole &str

Comment (WaffleLapkin) [reply library/core/src/str/mod.rs:1239]: Is there a use case when you need to do those kinds of tricks?

Comment (WaffleLapkin) [reply library/core/src/str/mod.rs:1239]: I mean the fix is trivial, but for me, it's not clear if we should provide such guarantees.

**Comment by WaffleLapkin** [comment]
Hm. I've fixed the conflict, but is it normal that bors didn't change the labels (`S-waiting-on-review`->`S-waiting-on-author`)? It said to run a command after resolving conflicts, but labels are already as they should be after resolving conflicts.

**Comment by Dylan-DPC** [comment]
@WaffleLapkin not sure why, the change has been reported last week as well, will check it accordingly :)

**Review by dtolnay — APPROVED** [review]
Thanks, this looks good!

Please make a tracking issue and update the stability attributes in the PR.

**Comment by WaffleLapkin** [comment]
@dtolnay done :)

**Comment by leonardo-m** [comment]
What's an use case of this as_str method?

**Comment by michaelvoronov** [comment]
@leonardo-m I needed this feature while implementing very simple argument parser.

Let's imagine that I want to parse the following command:
`call <module_name> <func_name> [args in json]`

So in `call` after `<func_name>` extracting, I want to get a remainder of string as `&str` to parse it like a json with `serde`. In particular, [here](https://github.com/fluencelabs/fce/blob/master/tools/repl/src/repl.rs#L107) I had to use `join` or `collect` that implies some overhead.

**Comment by crumblingstatue** [comment]
Would it be possible to also implement this for SplitWhitespace?

**Comment by WaffleLapkin** [comment]
@crumblingstatue this should be possible, but it's not trivial. `SplitWhitespace` [uses] `Filter` iterator adaptor, making it impossible to reach inner `Split` and get the remainder. The only solution I can think of - making `Filter`'s [`iter` field](https://doc.rust-lang.org/src/core/iter/adapters/mod.rs.html#953) `pub(crate)`.

[uses]: https://doc.rust-lang.org/nightly/src/core/str/iter.rs.html#1150

File: library/core/src/str/mod.rs (hunk 1 above)
Comment (loriopatrick) [reply library/core/src/str/mod.rs]: I would like the return to be a slice of the provided input.

When parsing a http header, I find myself wanting to see how much is remaining in the buffer to determine if I have all the contents (specified by Content-Length).

Example

```
let http_response_buffer: &str;
let mut header_lines = http_response_buffer.split("\r\n");

let mut content_length = None;
/* read through header and set content_length, header ends on empty header_line */
/* ... */

let unread_buffer_size = unsafe {
   header_lines.as_str().as_ptr().offset_from(response_buffer.as_ptr())
};

if content_length.unwrap() < unread_buffer_size {
   panic!("http_response_buffer does not contain response body");
}
```

Comment (WaffleLapkin) [reply library/core/src/str/mod.rs]: I was thinking about this today and thought that there should be a function (probably in a crate) like
```rust
fn substr_bounds(string: &str, sub: &str) -> Option<Range<usize>> { ... }

let string = "aabb";
let sub = &r[1..3];
assert_eq!(substr_bound(string, sub), Some(1..3));
assert_eq!(substr_bound(string, "ab"), None);
```

It feels natural, but are there any cases when it's useful? Anyway, as said before, for these kinds of things returning an empty substr may be a good idea.

If there is code that may benefit from this, I could make a PR with a fix.

Comment (HyeonuPark) [reply library/core/src/str/mod.rs]: It should be related to the semantics of raw pointer comparison.
https://github.com/rust-lang/unsafe-code-guidelines/issues/239
